Update docker provider documentation (#2804)
elastic/observability-docs: docs/en/ingest-management/elastic-agent/configuration/providers/docker-provider.asciidoc

[[docker-provider]]
= Docker Provider

Provides inventory information from Docker.

> Docker provider ensures, that each docker container event is enriched with
> the container's metadata, namely: `container.id`, `container.name`, `container.image.name`
> and `container.labels`.

Reviewer: ```suggestion
Docker provider ensures that each docker container event is enriched with
the container's metadata, namely: `container.id`, `container.name`, `container.image.name`,
and `container.labels`.

Just a super small nitpick for consistency with our style in other docs. :-) 
```

Sample of the policy generated by this configuration will look like:

[source,yaml]
----
inputs:
  - id: docker-container-logs-b9b898d9c2a1126384d38e9f857b3985480cd05c8e74ffc8b628d92245f5a103
    streams:
      paths:
      - /var/lib/docker/containers/b9b898d9c2a1126384d38e9f857b3985480cd05c8e74ffc8b628d92245f5a103/*-json.log
    processors:
    - add_fields:
        fields:
          id: b9b898d9c2a1126384d38e9f857b3985480cd05c8e74ffc8b628d92245f5a103
          image: image-name:latest
          labels:
            key: value
          name: container-name
        target: container
  - id: docker-container-596bbd114498253985e6a5c4f0f7bf2d9eb8fcd4fe3e6cb53bdfba0cdc7036c8
    type: filestream
    streams:
      paths:
      - /var/lib/docker/containers/596bbd114498253985e6a5c4f0f7bf2d9eb8fcd4fe3e6cb53bdfba0cdc7036c8/*-json.log
    processors:
    - add_fields:
        fields:
          id: 596bbd114498253985e6a5c4f0f7bf2d9eb8fcd4fe3e6cb53bdfba0cdc7036c8
          image: other-image-name:latest
          labels:
            key: value
          name: other-container-name
        target: container
----
> NOTE: Docker provider ensures that each docker container event is enriched with

> the container's metadata, for this inputs will be populetad with `add_fields` processor

Reviewer 2: ```suggestion
the container's metadata, and hence the inputs will be populated with the `add_fields` processor which will be responsible for adding the proper container's metadata.
```

|===
|Key |Type |Description

|`docker.container.id`
|`string`
|ID of the container

|`docker.container.name`
|`string`
|Name of the container

> |`docker.container.image.name`

Author: note: wait merging of https://github.com/elastic/elastic-agent/pull/2411 first

|===
|Key |Type |Description

|`docker.container.id`
|`string`
|ID of the container

|`docker.container.name`
|`string`
|Name of the container

|`docker.container.image.name`
|`string`
|Image of the container

|`docker.container.labels`
|`string`
|Labels of the container
|===
// lint enable arg

To set the container ID dynamically in the configuration, use a variable in the
{agent} policy to return container ID information from the provider:

[source,yaml]
----
inputs:
  - id: 'docker-container-logs-${docker.container.id}'
    type: filestream
    paths:
      - '/var/lib/docker/containers/${docker.container.id}/*-json.log'
----

Sample of the policy generated by this configuration will look like:

[source,yaml]
----
inputs:
  - id: docker-container-logs-b9b898d9c2a1126384d38e9f857b3985480cd05c8e74ffc8b628d92245f5a103
    paths:
    - /var/lib/docker/containers/b9b898d9c2a1126384d38e9f857b3985480cd05c8e74ffc8b628d92245f5a103/*-json.log

>   - id: docker-container-596bbd114498253985e6a5c4f0f7bf2d9eb8fcd4fe3e6cb53bdfba0cdc7036c8

Reviewer 2: Hmm, how about keeping the `processors` part in place?
```
processors:
      - add_fields:
          container.name: my-container
          container.labels: ....
          ...
```
It would help us understand what metadata will be added.

Author: adding those fields - is as well an implementation detail - since it defined [here](https://github.com/elastic/elastic-agent/blob/e7419ddc4e492e7e41acd20ec39f570937affe13/internal/pkg/composable/providers/docker/docker.go#L140-L153), and there is no control over this list

Reviewer 2: The point is that this would show how the events will be enriched with the container's metadata.
Do we have any other resource that would inform the users which metadata are added to the docker container logs?

Author: no, we don't, except the keys list. I've added this explanation - https://github.com/elastic/observability-docs/pull/2804/commits/4bdbb67976b5894261439a9f10eabbf8211ee0d2, do you think it is clear enough?

Reviewer 2: Commented there.

[
    {
       "id": "1",
       "mapping:": { "container": {"id": "1", "name": "container-name", "image": {"name": "busybox:latest"}}},
       "processors": {"add_fields": {"container.name": "my-container"}}
    },
    {
        "id": "2",

>         "mapping:": { "container": {"id": "2", "name": "container-name2", "image": {"name": "busybox:latest"}}},

Author: looking on this example again - do you think it makes sense to add it to the public doc?
if I correctly understood, this example references to the [implementation](https://github.com/elastic/elastic-agent/blob/e7419ddc4e492e7e41acd20ec39f570937affe13/internal/pkg/composable/providers/docker/docker.go#L29-L33), since we have documentation in different repo - it is easy to forget changing it, so it can be outdated. Also it is not smth that customer can configure, so maybe just keep the `inputs:` and `generated policy` examples here?
@ChrsMark wdyt?

>         "mapping:": { "container": {"id": "2", "name": "container-name2", "image": {"name": "busybox:latest"}}},

Reviewer 2: Yeap, good point. I agree we can remove this since it's not user facing behavior but mostly internal implementation detail.

>         "mapping:": { "container": {"id": "2", "name": "container-name2", "image": {"name": "busybox:latest"}}},

Author: done https://github.com/elastic/observability-docs/pull/2804/commits/0a56e4300240d3d03a3d6201afbf7781b388cb6c

        "processors": {"add_fields": {"container.name": "other-container"}}
    }
]
----

{agent} automatically prefixes the result with `docker.`, which is the name of the provider:

[source,json]
---
[
    {"docker": {"container": {"id": "1", "name": "container-name", "image": {"name": "busybox:latest"}}}},
    {"docker": {"container": {"id": "2", "name": "container-name2", "image": {"name": "busybox:latest"}}}},
]
---

To set the container ID dynamically in the configuration, use a variable in the
{agent} policy to return container ID information from the provider:

[source,yaml]
----
inputs:
  - id: 'docker-container-logs-${docker.container.id}'
    type: filestream
    paths:
      - '/var/lib/docker/containers/${docker.container.id}/*-json.log'
----

The policy generated by this configuration looks like:

[source,yaml]
----
inputs:
  - id: docker-container-logs-1
    paths:

>     - /var/lib/docker/containers/1/*-json.log

Reviewer 2: Maybe better to use a more meaningful ID example:
```suggestion
    - /var/lib/docker/containers/79f03945994b3f1c629c677796b1985593663f4b8304b29138a5524caae3ba25/*-json.log
```
